File: kangtaku/employee_importance.java
/*
// Employee info
class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};
*/
class Solution {
    Map<Integer, Integer> mp;
    List<Integer>[] gp;
    public int getImportance(List<Employee> employees, int id) {
        gp = new ArrayList[2001];
        mp = new HashMap<>();
        for (Employee emp : employees) {
            mp.put(emp.id, emp.importance);
            if (gp[emp.id] == null) gp[emp.id] = new ArrayList<>();
            for (Integer sub : emp.subordinates) {
                gp[emp.id].add(sub);
            }
        }
        
        return dfs(id);
    }
    
    public int dfs(int id) {
        int res = mp.get(id);
        for (Integer sub : gp[id]) {
            res += dfs(sub);
        }
        return res;
    }
}

File: iamminji/leetcode/easy/employee_importance.py
# 690. Employee Importance
# https://leetcode.com/problems/employee-importance/description/


"""
# Employee info
class Employee:
    def __init__(self, id, importance, subordinates):
        # It's the unique id of each node.
        # unique id of this employee
        self.id = id
        # the importance value of this employee
        self.importance = importance
        # the id of direct subordinates
        self.subordinates = subordinates
"""


class Solution:
    def helper(self, visited, subs, employees):
        res = 0
        if not subs:
            return res

        for sub_id in subs:
            if sub_id in visited:
                return visited[sub_id]

            for emp in employees:
                if emp.id == sub_id:
                    res += emp.importance + self.helper(visited, emp.subordinates, employees)
                    visited[sub_id] = res
        return res

    def getImportance(self, employees, id):
        """
        :type employees: Employee
        :type id: int
        :rtype: int
        """
        visited = dict()
        res = 0
        for emp in employees:
            if emp.id == id:
                visited[emp.id] = True
                res = emp.importance + self.helper(visited, emp.subordinates, employees)
        return res
